### DISCUSSION TITLE
[Bug]: XDebug not working on HerdPro with PHP Storm

### DISCUSSION THREAD

**Discussion by craigvc** [beyondcode/herd-community#435] [Q&A] [answered]
### Operating system version

Mac OS Sonoma 14.3.1 (M3)

### System architecture

ARM64 (M1, M2, etc)

### Herd Version

1.4.1

### PHP Version

8.2.16

### Bug description

I am trying to debug from phpstorm
I have herd pro.
I check my php-v and get this
➜  Herd php -v
PHP 8.2.16 (cli) (built: Feb 16 2024 00:30:02) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.2.16, Copyright (c) Zend Technologies
    with Zend OPcache v8.2.16, Copyright (c), by Zend Technologies
    
    so it seems xdebug isnt installed?
    how do i fix this?
    

### Steps to reproduce

Install PHP storm, install herd on MAC Ventura

**Comment by mpociot** [comment]
The CLI version will only activate Xdebug when you run "herd debug". 
You can use this as a "php" command replacement, like "herd debug -v"

**Comment by craigvc** [comment]
ah so i need the browser extension for debugging a site? I thought the pro version did it automatically?

**Comment by mpociot** [comment]
The pro version automatically enables Xdebug on-the-fly, either by setting a breakpoint in your PHPStorm project, or if you don't use PHPStorm, by toggling the Xdebug extension.

Be sure to save the file after setting a breakpoint, so that Herd can detect it.

**Comment by craigvc** [comment]
Maybe I am missing something and apologies if its dumb questions!
Ok so i set a breakpoint on phpstrom, listen for connections in phpstorm
Then visit herd site - it should just work yes?
Do i need to install a browser extension or something?

**Comment by mpociot** [comment]
Yeah, that's how it should work. You should be able to verify this by returning `phpinfo()`.
With a breakpoint, xdebug should be enabled.

One thing you might want to look out for: We detect your breakpoint by looking for a folder in your parked paths, that matches the hostname.
If you linked your site with a different name, this detection might not work and you need to use the extension instead.

**Comment by craigvc** [comment]
Something is not good so if i do phpinfo xdebug isnt in it?

**Comment by craigvc** [comment]
The paths are not changed from the site name btw

**Comment by mpociot** [comment]
Just to double check, if you use the Xdebug browser extension, it works as expected?

**Comment by craigvc** [comment]
no doesnt work there either?

**Comment by mpociot** [comment]
Weird. Could you please toggle the checkbox in the Debugger settings? Maybe that setting was not properly persisted

**Comment by craigvc** [comment]
ah getting somewhere!
It may be caused by path mappings misconfiguration or not synchronized local and remote projects.  To figure out the problem check path mappings configuration for 'gcc-v2.test' server at PHP|Servers or enable 'Break at first line in PHP scripts' option (from Run menu). Do not show again

**Comment by craigvc** [comment]
<img width="1882" alt="image" src="https://github.com/beyondcode/herd-community/assets/13494843/c8e339fe-8f02-47f0-88f3-6449f4dfb768">

**Comment by craigvc** [comment]
Any clues to this error?
I moved the project now to the Herd folder instead of my normal folder and its now throwing this error so its getting something from herd!

(with and without the chrome extension)

**Comment by marvinosswald** [comment]
> Weird. Could you please toggle the checkbox in the Debugger settings? Maybe that setting was not properly persisted

this fixed it for me 👍

**Comment by craigvc** [comment]
> > Weird. Could you please toggle the checkbox in the Debugger settings? Maybe that setting was not properly persisted
> 
> this fixed it for me 👍

I tried this several times

**Comment by mpociot** [comment]
@craigvc what happens when you enable "break at first line" in PHPStorm, as the error message suggests?

**Comment by craigvc** [comment]
it brings me to this page in php storm dump-loader.php
```

use BeyondCode\HerdConfiguration\HerdConfiguration;

/*
|--------------------------------------------------------------------------
| Laravel Herd
|--------------------------------------------------------------------------
|
| This script is the entrypoint of your Laravel Herd request. If you see
| this file while using Herd's Xdebug auto-detection, you may want to
| modify your PHPStorm settings, to not automatically break at the
| first line.
|
| Learn more about it in the Herd documentation
| https://herd.laravel.com/docs/1/herd-pro/xdebug
|
*/

require_once __DIR__ . '/HerdConfiguration.php';

try {
    $config = HerdConfiguration::load();
} catch (\Throwable $e) {
    return;
}
```

**Comment by philipmclifton** [comment]
I am also having this issue again after originally finding success with the break at first breakpoint. dumping phpinfo shows that Xdebug is not loaded in the request at all even when breakpoints exist.  and checking the workspace.yml file to confirm the following exists:

```
 <component name="XDebuggerManager">
    <breakpoint-manager>
      <breakpoints>
        <line-breakpoint enabled="true" type="php">
          <url>file://$PROJECT_DIR$/app/Traits/GiftCards/Orders/InteractsWithTemporaryOrderRows.php</url>
          <line>11</line>
          <option name="timeStamp" value="1" />
        </line-breakpoint>
        <line-breakpoint enabled="true" type="php">
          <url>file://$PROJECT_DIR$/app/Traits/GiftCards/Orders/InteractsWithTemporaryOrderRows.php</url>
          <line>9</line>
          <option name="timeStamp" value="2" />
        </line-breakpoint>
        <line-breakpoint enabled="true" type="php">
          <url>file://$PROJECT_DIR$/app/Traits/GiftCards/Orders/InteractsWithTemporaryOrderRows.php</url>
          <line>13</line>
          <option name="timeStamp" value="3" />
        </line-breakpoint>
      </breakpoints>
    </breakpoint-manager>
  </component>
```


<img width="849" alt="image" src="https://github.com/beyondcode/herd-community/assets/3420435/660c6a0f-63db-4660-b9c7-164dc0c77fcd">

**Comment by philipmclifton** [comment]
Ok - Toggling the checkbox in the herd debug settings page worked for me this time!

**Comment by craigvc** [comment]
@mpociot I am using filamentphp and livewire could that have anything do with it?
I see like 122 notifications about the Debug session was finished without being paused when i reload a page in the browser (I see it in PHPStorm)

**Comment by paulmichaeldev** [comment]
> it brings me to this page in php storm dump-loader.php
> 
> ```
> 
> use BeyondCode\HerdConfiguration\HerdConfiguration;
> 
> /*
> |--------------------------------------------------------------------------
> | Laravel Herd
> |--------------------------------------------------------------------------
> |
> | This script is the entrypoint of your Laravel Herd request. If you see
> | this file while using Herd's Xdebug auto-detection, you may want to
> | modify your PHPStorm settings, to not automatically break at the
> | first line.
> |
> | Learn more about it in the Herd documentation
> | https://herd.laravel.com/docs/1/herd-pro/xdebug
> |
> */
> 
> require_once __DIR__ . '/HerdConfiguration.php';
> 
> try {
>     $config = HerdConfiguration::load();
> } catch (\Throwable $e) {
>     return;
> }
> ```

I'm also experiencing this - I am able to step through it to my actual breakpoint, but I get the feeling there's an option in PHPStorm that is a little wonky.

Every request served by PHP gets halted at line 20 in /Applications/Herd.app/Contents/Resources/valet/dump-loader.php.

My project is Vue with Inertia, so I _think_ that might rule out filament/livewire @craigvc

**Comment by ju5t** [comment]
I'm using Livewire but unfortunately xdebug is not working for me either. It keeps breaking at the dump-loader script. Break at first line in script is disabled.

**Comment by craigvc** [comment]
@mpociot can we get a little guidance here ?

**Comment by craigvc** [comment]
@paulmichaeldev I did read that can actually be normal in the sense its sending lots of ajax requests - so would apply from inertia and livewire perspective. Issue i am stuck at is i cant even step through or it doesnt detect my breakpoints - phpstorm EAP and 2023.3.4

**Comment by mpociot** [comment]
@craigvc if it stops for you in the dump-loader.php file, this is the default behavior of PHPStorm as it breaks on the first line.
You should be able to just continue debugging / running the script using the PHPStorm debugger. At this point, xdebug is already activated

**Reply by ju5t** [reply]
I've tried this with Livewire, but that doesn't work. Even if I skip through things quickly with fn+F8, it takes ages before it's gone through the Valet files and before I can reach my component xdebug stops and the process continues in my browser.

Perhaps we can't use xdebug with Livewire? What specific information would you need from me to debug this further?

**Reply by nathanheffley** [reply]
@ju5t why are you using fn+F8? That just steps through the program, is Resume Program not stopping at your breakpoint? Also, check to see if "Force break at first line when a script is outside the project" is turned off. The dump-loader is not technically in your project and it was turned on and causing problems for me by default. It's on the same screen as the option to break at the first line.

**Comment by paulknisely** [comment]
I am also running into the issue with it stopping in the dump-loader.php file. If I don't have it set to break at first line in PHP scripts, it never pops up the debugger. If I have that setting checked, it starts in the dump-loader.php file and I have to click so many times that I often click past my break point.

The cli debugging is flawless. 'herd debug artisan' works as described and is great.

I'm sure it's a skill issue on my part but at the moment, xdebug through the browser is unusable. I've also never had xdebug working but now that I see more and more people talk about the benefits, I feel like I'm missing out on a great tool.

I'm more than willing to help, I'm just unsure where and how to help.

**Comment by philipmclifton** [comment]
Would it be possible to include a switch to force xdebug to be on or off as a fallback?

**Comment by craigvc** [comment]
I dont believe you should have closed this as it is still a bug.
I just wiped my mac m1 Pro, clean install - herd, phpstorm - just trial version to eliminate settings sync DBEngine and Arc Browser - that was it.
ran phpstorm started listening and hit the dumploader file - trying to debug just loops it it doesnt work.
It is a filamentphp project. @mpociot

**Comment by nathanheffley** [comment, marked as answer]
I noticed that even with "Break at first line in PHP scripts" turned off, it would still break in the dump-loader. Then I noticed "Force break at first line when a script is outside the project" and after turning that off, because the dump-loader is in the vendor directory which is not considered to be in the project, it now only stops on my breakpoints.

I have noticed that auto-saving is not good enough for Herd to detect the breakpoints. Cmd+S after changing a breakpoint is still necessary.

Hopefully this helps a few people here.

**Reply by craigvc** [reply]
Spot on with those tips sir!

**Reply by jonlod** [reply]
Thanks for the tips, worked like a charm! This should be noted in the dump-loader file, not the vague redirect that's there now: 
`This script is the entrypoint of your Laravel Herd request. If you see this file while using Herd's Xdebug auto-detection, you may want to modify your PHPStorm settings, to not automatically break at the first line.`

**Comment by philipmclifton** [comment]
Anyone found a way to enable xdebug when running tests inside phpstorm? Xdebug works in the browser and when using herd "command" but please see screenshot
<img width="1223" alt="Screenshot 2024-04-10 at 15 39 30" src="https://github.com/beyondcode/herd-community/assets/3420435/4a35adcd-98a7-4ab3-9a63-10a95bf5a7ec">

I still think we need a force enable xdebug toggle

**Reply by nathanheffley** [reply]
Does putting a breakpoint in some dead code just to toggle Herd's Xdebug make it work?
I agree some better way to see the status of Xdebug and manually override it would be nice.

**Comment by larsclosinglock** [comment]
Same issue here. Installing Herd still breaks debugging in PHPStorm.

**Reply by nathanheffley** [reply]
What's broken about it? OP's problem seems to be resolved with the steps in my selected answer comment.

**Comment by inxilpro** [comment]
I was having this issue, too. My solution:

1. Open up Herd's config (Herd menu > Open configuration files)
2. Open the `debug/debug.ini` inside your PHP version
3. Copy the path to the `zend_extension` referenced in that file (ie. `/Applications/Herd.app/Contents/Resources/xdebug/xdebug-81-arm64.so`)
4. In PhpStorm, go to Settings, then PHP, and click the "..." button next to your CLI interpreter configuration
5. Paste the zend extension path into the "Debugger extension" option in your interpreter config

This will then cause PhpStorm to load that extension, regardless of whether auto-detect is working, whenever you run a test in debug mode.

Hope that helps!

**Comment by spencerstith** [comment]
Hello, this is still breaking for me!
I have xdebug installed (using the included `.so` file with Herd)  and PhpStorm is able to use the debugger now when running scratch files (couldn't do before), but I am unable to hit the breakpoint when using the browser.

When I first set it all up, I did hit a breakpoint in some random code when using the browser, outside of my project. But now, not even that happens any longer and i can't hit any breakpoints outside of the scratch files that PhpStorm runs itself.

If it's any help, when i do finish a debug session in a scratch file, PhpStorm warns
'Cannot load Xdebug - it was already loaded'. idk if that really means much though.

I have the following contents in my PHP ini:

```
curl.cainfo=/Users/spencerstith/Library/Application Support/Herd/config/php/cacert.pem
openssl.cafile=/Users/spencerstith/Library/Application Support/Herd/config/php/cacert.pem
pcre.jit=0


memory_limit=128M
upload_max_filesize=2M

zend_extension=/Applications/Herd.app/Contents/Resources/xdebug/xdebug-82-arm64.so
xdebug.mode=debug,develop
xdebug.start_with_request=yes
xdebug.start_upon_error=yes
```

I get no errors in my PHP logs or in the browser console.

If anyone could help that would be awesome! I have tried a few of the solutions listed on this discussion but none of them seemed to fix the issue.

**Reply by ahuininga-orisha** [reply]
Wat worked for me was removing this from the php.ini, and removing the servers in phpstorm under PHP->servers. And making sure my breakpoint is saved by adding an line and saving the file after adding the breakpoint.

```
zend_extension=/Applications/Herd.app/Contents/Resources/xdebug/xdebug-82-arm64.so
xdebug.mode=debug,develop
xdebug.start_with_request=yes
xdebug.start_upon_error=yes
```

**Comment by sheadawson** [comment]
Toggling "Detect breakpoints automatically" in herd debugger settings fixed this for me

**Comment by darwin-buelo** [comment]
Does any able to run Herd pro Xdebug in PHPstrom. on windows?

**Reply by johannesobier** [reply]
Yes, I use Herd Pro. The reason why it didn't work before was silly: I didn't remember that long time ago I added the OS environment variable PHP_IDE_CONFIG=serverName={some server name here}. After a lot of tries I saw PhpStorm was showing a hint in the debug window "Threads & Variables" about it. After removing the OS environment variable and setting back PhpStorm's debug configuration to reasonable settings, now it works like a charm.
